internal/server/cache/cache.go

```diff
@@ -16,30 +16,78 @@ package cache
 
 import (
 	"context"
+	"database/sql"
 	"encoding/json"
 	"log"
 	"os"
+	"sync"
 
+	pb "github.com/datacommonsorg/mixer/internal/proto"
 	"github.com/datacommonsorg/mixer/internal/server/resource"
-	"github.com/datacommonsorg/mixer/internal/server/statvar"
+	"github.com/datacommonsorg/mixer/internal/server/statvar/fetcher"
+	"github.com/datacommonsorg/mixer/internal/server/statvar/hierarchy"
+	"github.com/datacommonsorg/mixer/internal/sqldb/query"
 	"github.com/datacommonsorg/mixer/internal/store"
 )
 
 const (
 	blockListSvgJsonPath = "/datacommons/svg/blocklist_svg.json"
 )
 
-type SearchOptions struct {
-	UseSearch           bool
-	BuildSvgSearchIndex bool
+// Options for using the Cache object
+type Options struct {
+	FetchSVG   bool
+	SearchSVG  bool
+	CustomProv bool
 }
 
-// NewCache initializes the cache for stat var hierarchy.
-func NewCache(
-	ctx context.Context,
-	store *store.Store,
-	searchOptions SearchOptions,
-) (*resource.Cache, error) {
+// Cache holds cached data for the mixer server.
+type Cache struct {
+	// ParentSvg is a map of sv/svg id to a list of its parent svgs sorted alphabetically.
+	parentSvg map[string][]string
+	// SvgInfo is a map of svg id to its information.
+	rawSvg map[string]*pb.StatVarGroupNode
+	// A list of blocked top level svg.
+	blockListSvg map[string]struct{}
+	// SVG search index
+	svgSearchIndex *resource.SearchIndex
+	// Custom provenance from SQL storage
+	customProvenances map[string]*pb.Facet
+	// Lock for updating cache
+	mu sync.RWMutex
+}
+
+func (cache *Cache) ParentSvg() map[string][]string {
+	cache.mu.RLock()
```

Comment:
One approach for avoiding read locks completely is maintaining cached data in a separate structure, say `CachedData` and simply swapping it on cache refresh. You can still maintain a single lock for when you swap but should not need it otherwise.

```
struct CachedData {
  parentSvgs map[string][]string
  ...
}

struct Cache {
  lock sync.Mutex
  data CachedData
}
```